openwisp_controller/config/whois/tasks.py

```diff
@@ -158,3 +167,93 @@ def delete_whois_record(ip_address):
     queryset = WHOISInfo.objects.filter(ip_address=ip_address)
     if queryset.exists():
         queryset.delete()
+
+
+@shared_task
+def manage_fuzzy_locations(
+    device_pk,
+    ip_address,
+    latitude=None,
+    longitude=None,
+    address=None,
+    add_existing=False,
+):
+    """
+    Creates/updates fuzzy location for a device based on the latitude and longitude
+    or attaches an existing location if `add_existing` is True.
+    Existing location here means a location of another device whose last_ip matches
+    the given ip_address.
+
+    - If attaching existing location and current device has no location then the
+    existing location is attached to the current device as well.
+    - If attaching existing location and current device already has location, then
+    current device's location is deleted and set same as existing device's location.
+    - If not attaching existing then new location is created if no location exists for
+    current device, or existing one is updated if it is approximate(fuzzy).
+    """
+    Device = load_model("config", "Device")
+    Location = load_model("geo", "Location")
+    DeviceLocation = load_model("geo", "DeviceLocation")
+
+    device_location = (
+        DeviceLocation.objects.filter(content_object_id=device_pk)
+        .select_related("location")
+        .first()
+    )
+
+    if not device_location:
+        device_location = DeviceLocation(content_object_id=device_pk)
+
+    # if attaching an existing location, the current device should not have
+    # a location already set.
+    current_location = device_location.location
+    if add_existing and (not current_location or current_location.is_approximate):
+        existing_device_location = (
+            Device.objects.select_related("devicelocation")
```

Comment:
Should we filter to the current organization’s devices first, rather than checking every device?